### PR TITLE
[Merged by Bors] - REMOVE unsound lifetime annotations on `EntityMut`

### LINKED ISSUE #3408: World entity access is not sound
## Bevy version

0.5

## Operating system & version

Windows 10

## What you did

The World entity api allows use after free

```rust
use bevy::prelude::*;

struct Name(String);

fn test(world: &mut World) {
    let mut entity = world.spawn();
    entity.insert(Name("Test".to_string()));
    let name = entity.get::<Name>().unwrap();    
    entity.despawn(); // Or entity.remove::<Name>();
    println!("{}", name.0); // This line should be a compile error
}

fn main() {
    App::build()
        .add_system(test.exclusive_system())
        .run();
}
```

## What you expected to happen

The code should not compile

## What actually happened

The code compiles without any errors or warnings

### PR DESCRIPTION
**Pull request by BoxyUwU**
Fixes #3408
#3001 also solves this but I dont see it getting merged any time soon so...
# Objective
make bevy ecs a lil bit less unsound

## Solution
make `EntityMut::get_component_mut` return borrows from self instead of `'w`

### CODE REVIEW

**Review by Guvante — APPROVED** [review]

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 1]
```diff
@@ -62,17 +62,26 @@ impl<'w> EntityRef<'w> {
...
+    pub fn get<T: Component>(&self) -> Option<&'_ T> {
+        // SAFE: lifetimes enforce correct usage of returned borrow
+        unsafe { self.get_unchecked::<T>() }
+    }
+
+    /// This allows aliased mutability use after free bugs.
+    /// # Safety
+    /// - Must not be a mutable reference to the component.
+    /// - Must not use the returned reference after the component is removed
+    #[inline]
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/entity_ref.rs:74]:
```suggestion
    /// Gets the component of type `T` associated with this entity without checking for unique borrows.
    ///
    /// # Safety
    /// - Do not attempt to construct a mutable reference to the component while other references to this component are active or vice versa.
    /// - You must not use the returned reference after the component is removed
    #[inline]
```

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 2]
```diff
@@ -62,17 +62,26 @@ impl<'w> EntityRef<'w> {
...
+    pub unsafe fn get_unchecked<T: Component>(&self) -> Option<&'w T> {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/entity_ref.rs:75]: This is way too risky to be pub IMO.

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 3]
```diff
@@ -62,17 +62,26 @@ impl<'w> EntityRef<'w> {
...
         // SAFE: entity location is valid and returned component is of type T
-        unsafe {
-            get_component_with_type(self.world, TypeId::of::<T>(), self.entity, self.location)
-                .map(|value| &*value.cast::<T>())
-        }
+        get_component_with_type(self.world, TypeId::of::<T>(), self.entity, self.location)
+            .map(|value| &*value.cast::<T>())
     }
 
+    /// This allows aliased mutability and use after free bugs.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/entity_ref.rs:81]:
```suggestion
    /// Gets a mutable reference to the component of type `T` associated with this entity without checking for borrow uniqueness.
    ///
    /// # Safety
    /// - Do not construct this reference while another reference to the same component is active.
    /// - You must not use the returned reference after the component is removed.
```

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 4]
```diff
@@ -145,39 +154,32 @@ impl<'w> EntityMut<'w> {
...
-                    change_tick: self.world.change_tick(),
-                },
-            })
-        }
+    pub fn get_mut<T: Component>(&mut self) -> Option<Mut<'_, T>> {
+        // SAFE: world access is unique, and lifetimes enforce correct usage of returned borrow
+        unsafe { self.get_unchecked_mut::<T>() }
+    }
+
+    /// This allows aliased mutability use after free bugs.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/entity_ref.rs:168]: Same doc improvements as above...

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 5]
```diff
@@ -145,39 +154,32 @@ impl<'w> EntityMut<'w> {
...
+    /// - Must not be a mutable reference to the component.
+    /// - Must not use the returned reference after the component is removed
+    #[inline]
+    pub unsafe fn get_unchecked<T: Component>(&self) -> Option<&'w T> {
+        // SAFE: entity location is valid and returned component is of type T
+        get_component_with_type(self.world, TypeId::of::<T>(), self.entity, self.location)
+            .map(|value| &*value.cast::<T>())
     }
 
+    /// This allows aliased mutability and use after free bugs.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/entity_ref.rs:179]: More doc improvements.

**Review by alice-i-cecile — APPROVED** [review]
The UB is clear, and upon examining this more closely, the 'w lifetime is clearly incorrect. I've left some comments to improve the docs and safety comments.

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 6]
```diff
@@ -145,39 +162,40 @@ impl<'w> EntityMut<'w> {
...
-        }
+    pub fn get_mut<T: Component>(&mut self) -> Option<Mut<'_, T>> {
+        // SAFE: world access is unique, and lifetimes enforce correct usage of returned borrow
+        unsafe { self.get_unchecked_mut::<T>() }
+    }
+
+    /// Gets an immutable reference to the component of type `T` associated with
+    /// this entity without ensuring there are no unique borrows active and without
+    /// ensuring that the returned reference will stay valid.
+    /// # Safety
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/entity_ref.rs:179]:
```suggestion
    ///
    /// # Safety
    ///
```

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 7]
```diff
@@ -145,39 +162,40 @@ impl<'w> EntityMut<'w> {
...
+    pub unsafe fn get_unchecked<T: Component>(&self) -> Option<&'w T> {
+        // SAFE: entity location is valid and returned component is of type T
+        get_component_with_type(self.world, TypeId::of::<T>(), self.entity, self.location)
+            .map(|value| &*value.cast::<T>())
     }
 
+    /// Gets a mutable reference to the component of type `T` associated with
+    /// this entity without ensuring there are no other borrows active and without
+    /// ensuring that the returned reference will stay valid.
     /// # Safety
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/entity_ref.rs:194]:
```suggestion
    ///
    /// # Safety
    ///
```

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 8]
```diff
@@ -62,17 +62,34 @@ impl<'w> EntityRef<'w> {
...
-                .map(|value| &*value.cast::<T>())
-        }
+        get_component_with_type(self.world, TypeId::of::<T>(), self.entity, self.location)
+            .map(|value| &*value.cast::<T>())
     }
 
+    /// Gets a mutable reference to the component of type `T` associated with
+    /// this entity without ensuring there are no other borrows active and without
+    /// ensuring that the returned reference will stay valid.
     /// # Safety
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/entity_ref.rs:88]:
```suggestion
    ///
    /// # Safety
    ///
```

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 9]
```diff
@@ -62,17 +62,34 @@ impl<'w> EntityRef<'w> {
...
-    pub fn get<T: Component>(&self) -> Option<&'w T> {
+    pub fn get<T: Component>(&self) -> Option<&'_ T> {
+        // SAFE: lifetimes enforce correct usage of returned borrow
+        unsafe { self.get_unchecked::<T>() }
+    }
+
+    /// Gets an immutable reference to the component of type `T` associated with
+    /// this entity without ensuring there are no unique borrows active and without
+    /// ensuring that the returned reference will stay valid.
+    /// # Safety
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/entity_ref.rs:73]:
```suggestion
    ///
    /// # Safety
    ///
```

**Comment by alice-i-cecile** [comment]
@BoxyUwU can you resolve conversations now? :)

**Review by TheRawMeatball — APPROVED** [review]

**Pull request closed without merging**